graph/path/negative_cycles_example_test.go

```diff
@@ -60,3 +57,75 @@ func ExampleBellmanFordFrom_negativecycles() {
 	// negative cycle in path to c path:[c a b c]
 	// negative cycle in path to f path:[a b c a f]
 }
+
+func ExampleFloydWarshall_negativecycles() {
+	// FloydWarshall can be used to find an exhaustive
+	// set of nodes in negative cycles in a graph.
+
+	// Construct a graph with a negative cycle.
+	edges := []simple.WeightedEdge{
+		{F: simple.Node('a'), T: simple.Node('f'), W: -1},
+		{F: simple.Node('b'), T: simple.Node('a'), W: 1},
+		{F: simple.Node('b'), T: simple.Node('c'), W: -1},
+		{F: simple.Node('b'), T: simple.Node('d'), W: 1},
+		{F: simple.Node('c'), T: simple.Node('b'), W: 0},
+		{F: simple.Node('e'), T: simple.Node('a'), W: 1},
+		{F: simple.Node('f'), T: simple.Node('e'), W: -1},
+	}
+	g := simple.NewWeightedDirectedGraph(0, math.Inf(1))
+	for _, e := range edges {
+		g.SetWeightedEdge(e)
+	}
+
+	// Find the shortest path to each node from Q.
+	pt, ok := path.FloydWarshall(g)
+	if ok {
+		fmt.Println("no negative cycle present")
+		return
+	}
+
+	ids := []int64{'a', 'b', 'c', 'd', 'e', 'f'}
+
+	for _, id := range ids {
+		if math.IsNaN(pt.Weight(id, id)) {
+			fmt.Printf("%c is in a negative cycle\n", id)
+		}
+	}
+
+	for _, uid := range ids {
+		for _, vid := range ids {
+			_, w, unique := pt.Between(uid, vid)
+			if math.IsNaN(w) {
+				fmt.Printf("negative cycle in path from %c to %c unique=%t\n", uid, vid, unique)
+			}
+		}
+	}
+
+	// Output:
+	// a is in a negative cycle
+	// b is in a negative cycle
+	// c is in a negative cycle
+	// e is in a negative cycle
+	// f is in a negative cycle
+	// negative cycle in path from a to a unique=false
+	// negative cycle in path from a to e unique=false
+	// negative cycle in path from a to f unique=false
+	// negative cycle in path from b to a unique=false
+	// negative cycle in path from b to b unique=false
+	// negative cycle in path from b to c unique=false
+	// negative cycle in path from b to d unique=false
```

Comment:
There is a direct edge from b to d, its weight is 1, d is not reachable in other way, so why is this in the output? Am I missing something?

Reply:
Hee. This is the joy of negative cycles.

There is a direct path from b->d of cost 1, but there are n paths (b->c->b)_n->d of cost 1-2n. We don't restrict paths to be simple (the b->d path), because if we do and we still retain polynomial time (as we do here), we prove P=NP by example. And we don't want to do that.

Reply:
Oh.